cpp/src/parquet/encryption/crypto_factory.cc

```diff
@@ -172,8 +201,8 @@ std::shared_ptr<FileDecryptionProperties> CryptoFactory::GetFileDecryptionProper
     const KmsConnectionConfig& kms_connection_config,
     const DecryptionConfiguration& decryption_config, const std::string& file_path,
     const std::shared_ptr<::arrow::fs::FileSystem>& file_system) {
-  auto key_retriever = std::make_shared<FileKeyUnwrapper>(
-      &key_toolkit_, kms_connection_config, decryption_config.cache_lifetime_seconds,
+  auto key_retriever = std::make_shared<CryptoFactoryFileKeyRetriever>(
```

Comment:
It looks like we hit the issue of member variable lifecycle in the FileDecryptionProperties class. I see there is already a `FileDecryptionProperties::DeepClone()` function. Should we expect the end user to explicitly clone it?

Reply:
I don't follow what you're suggesting here sorry. I think the usual behaviour would be for users to generate new `FileDecryptionProperties` per-file, especially if external key material is used, as in that case the `FileKeyUnwrapper` holds a `FileKeyMaterialStore` that is file specific.

But from from some brief testing, it looks like when internal key material is used, it is possible to reuse `FileDecryptionProperties` that have been obtained from a `CryptoFactory` with multiple files. The key encryption ids are randomly generated, so they are different between files and don't conflict in the cache stored in the `KeyToolkit`. It also looks like calling `DeepClone` is unnecessary, as the decryption properties use a `DecryptionKeyRetriever` rather than using the the `column_decryption_properties_` and `footer_key_` fields that get wiped after a read.

Reply:
Thanks for the explanation! Could you point me to a real world code or test case having this issue if possible?

Reply:
My comment might have been a bit unclear, I didn't mean to suggest that it's an issue that the decryption properties can be reused without cloning first. The comment here might clarify things:
https://github.com/apache/arrow/blob/bd3fab4333f9e95680f5ed0cd931455e323e6e27/cpp/src/parquet/encryption/encryption.h#L365-L371

When using `FileDecryptionProperties` from a `CryptoFactory`, the condition that "this object keeps the keyRetrieval callback only, and no explicit keys or aadPrefix" is met, so using `DeepClone` isn't necessary and one instance of `FileDecryptionProperties` can be used for multiple files, but I don't think that's a problem.

Reply:
CMIW, did you mean that in this case users cannot have access to the CryptoFactory instance? In the parquet dataset implementation, we always keep all the required instances in the config: https://github.com/apache/arrow/blob/main/cpp/src/arrow/dataset/parquet_encryption_config.h#L63

Reply:
No this isn't related to having access to the CryptoFactory instance, I think this discussion has gotten a bit sidetracked from the main point of the PR.

My use case is probably a bit non-standard as I'm trying to provide a .NET binding to the encryption API without users having to deal with manual memory management. I have no control over the CryptoFactory lifetime so if users of ParquetSharp misuse the API they could get a segfault, which isn't usually expected in a managed language/runtime.

This got me thinking, I wonder how pyarrow handles this. And it turns out pyarrow has the same issue. This code causes a hang on my machine, but works if I don't have the `del crypto_factory` line:

```python
import base64
import gc
import pyarrow as pa
import pyarrow.parquet as pq
import pyarrow.parquet.encryption as pe


class TestKmsClient(pe.KmsClient):
    """ Dummy KMS client that base64 encodes keys """

    def __init__(self, kms_connection_configuration):
        pe.KmsClient.__init__(self)

    def wrap_key(self, key_bytes, master_key_identifier):
        return f"{base64.b64encode(key_bytes).decode('utf-8')}"

    def unwrap_key(self, wrapped_key, master_key_identifier):
        return base64.b64decode(wrapped_key)


# Create test data
x = pa.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], type=pa.int32())
y = pa.array([0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9], type=pa.float32())
schema = pa.schema([
    pa.field("x", pa.int32(), nullable=False),
    pa.field("y", pa.float32(), nullable=False),
])
table = pa.table([x, y], schema=schema)


# Write encrypted file
crypto_factory = pe.CryptoFactory(lambda config: TestKmsClient(config))
config = pe.KmsConnectionConfig()
encryption_config = pe.EncryptionConfiguration(
   footer_key="Key0",
   column_keys={
      "Key0": ["x", "y"],
   },
   double_wrapping=True,
   internal_key_material=True
)
encryption_properties = crypto_factory.file_encryption_properties(config, encryption_config)

with pq.ParquetWriter("./encrypted_py.parquet", table.schema,
                     encryption_properties=encryption_properties) as writer:
   writer.write_table(table)

# Read encrypted file
decryption_config = pe.DecryptionConfiguration()
decryption_properties = crypto_factory.file_decryption_properties(config, decryption_config)

del crypto_factory
gc.collect()

with pq.ParquetFile("./encrypted_py.parquet",
                     decryption_properties=decryption_properties) as reader:
   read_table = reader.read()

print(read_table)
```

A similar example using a more realistic encrypted file caused a segfault.

Reply:
> My use case is probably a bit non-standard as I'm trying to provide a .NET binding to the encryption API without users having to deal with manual memory management.

This is similar to the java binding via JNI where the lifetime management is tricky. Usually we provide a `releaseXXX` native call for users to explicitly release the resource managed by the native code. I'm not sure if you can do the similar thing.

> This got me thinking, I wonder how pyarrow handles this. And it turns out pyarrow has the same issue.

pyarrow users have access to the crypto_factory, so I don't think this is a problem.

Reply:
Sorry for bearing with me. Now I think I have fully understood your case.

The patch itself LGTM. Though I'm not fully convinced this is a bug rather than an unexpected use case.

cc @pitrou @emkornfield @mapleFU for more opinions.

Reply:
Yes, users can work around this by holding a reference to the CryptoFactory. But I think it would be good to improve the API here so that users didn't have to worry about this, especially for users of managed languages/runtimes like C# and Python where they don't expect usage errors to lead to segfaults or undefined behaviour.

It's fair to say this isn't exactly a bug, but I'd say that if this behaviour isn't changed then at the very least the Python API documentation for `CryptoFactory.file_decryption_properties` should be updated to warn users about the possibility of undefined behaviour if they don't keep their CryptoFactory alive as long as the decryption properties may be in use.

Reply:
Thanks for the clarification @adamreeve and thanks for the ping @wgtmac .

1. I agree that we should avoid unowning pointers. as far as possible, except where performance-critical (which isn't the case here).
2. If this fixes a crash in Python, then we should also add a Python test. Can you do that @adamreeve ?

Reply:
Yes sure, I will add a Python test for this.

Reply:
I've added a Python test for this now